nix/onchain.nix

```diff
@@ -0,0 +1,534 @@
+                        | xargs fourmolu ${extStr} -m check
+                  '';
+              };
+
+            cabalFormatCheck =
+              lib.ifEnable projectConfig.enableCabalFormatCheck
+                {
+                  cabalFormatCheck = utils.shellCheck
+                    "cabalFormatCheck"
+                    projectConfig.src
+                    {
+                      nativeBuildInputs = [ cabalFmt ];
+                    }
+                    ''
+                      find -name '*.cabal' -not -path './dist*/*' -not -path './haddock/*' | xargs cabal-fmt -c
+                    '';
+                };
+
+            checks =
+              lib.fold lib.mergeAttrs { }
+                [
+                  flake.checks
+                  buildChecks
+                  haskellFormatCheck
+                  cabalFormatCheck
+                ];
+
+            haskellSources = "$(git ls-tree -r HEAD --full-tree --name-only | grep '.\\+\\.hs')";
+            cabalSources = "$(git ls-tree -r HEAD --full-tree --name-only | grep '.\\+\\.cabal')";
+          in
+          {
+            devShell = flake.devShell;
+
+            inherit checks;
+
+            check = utils.combineChecks "combined-checks" checks;
+
+            run.nixFormat =
+              {
+                dependencies = [ nixpkgsFmt ];
+                script = ''
+                  find . -name '*.nix' -not -path './dist*/*' -not -path './haddock/*' -exec nixpkgs-fmt {} +
+                '';
+                groups = [ "format" "precommit" ];
+                help = ''
+                  echo "  Formats nix files using nixpkgs-fmt."
+                '';
+              };
+            run.haskellFormat =
+              let
+                arguments = builtins.concatStringsSep " " (builtins.map (extension: "-o -X" + extension) projectConfig.ghc.extensions);
+              in
+              {
+                dependencies = [ fourmolu cabalFmt ];
+                script = ''
+                  # shellcheck disable=SC2046
+                  fourmolu ${arguments} -m inplace ${haskellSources}
+                  # shellcheck disable=SC2046
+                  cabal-fmt -i ${cabalSources}
+                '';
+                groups = [ "format" "precommit" ];
+                help = ''
+                  echo "  Runs fourmolu and cabal-fmt."
+                  echo
+                  echo "  fourmolu: A formatter for Haskell source code."
+                  echo "  cabal-fmt: Format .cabal files preserving the original field ordering, and comments."
+                  echo
+                  echo "  Fourmolu is using the following Haskell extensions:"
+                  echo "${builtins.concatStringsSep "\n" (builtins.map (p: "  - " + p) projectConfig.ghc.extensions)}"
+                  echo
+                  echo "  NOTE: You can change these in the flake module!"
+                '';
+              };
+            run.haskellLint =
```

Comment:
What about in-place refactoring? Also, refactoring takes a long time -- can we can also do a `refactor_changed` such as currently exists on `liqwid-onchain`

Reply:
I'd prefer having separate script for in-place and non-in-place as sometimes linter messes things up.

Reply:
Acutally, I think it will be better to make this accept argument. So when you call `nix run .#haskellLint` it just checks but if you call `nix run .#haskellLint -- apply` is does the inplace lint

Reply:
That's strange. Can you give an example of the error message you're seeing?

Reply:
it just said it has no permission. I closed the terminal, so I don't have it now.

Reply:
I'm binning this feature as something for a later version. It's definitely something we could want, but it's not critical enough to block our initial version release.